cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/AbstractArtifactLocalizer.java

```diff
@@ -0,0 +1,273 @@
+      throw new RetryableException(String.format("The response from %s did not contain the %s header.",
+                                                 urlConn.getURL(), HttpHeaders.LAST_MODIFIED));
+    }
+
+    return lastModified;
+  }
+
+  /**
+   * Helper method for catching and throwing any errors that might have occurred when attempting to connect.
+   */
+  private void throwIfError(HttpURLConnection urlConn, ArtifactId artifactId)
+    throws IOException, ArtifactNotFoundException {
+    int responseCode = urlConn.getResponseCode();
+    if (responseCode == HttpURLConnection.HTTP_OK) {
+      return;
+    }
+
+    String errMsg = CharStreams.toString(new InputStreamReader(urlConn.getErrorStream(), StandardCharsets.UTF_8));
+    switch (responseCode) {
+      case HttpURLConnection.HTTP_NOT_FOUND:
+        throw new ArtifactNotFoundException(artifactId);
+      case HttpURLConnection.HTTP_UNAVAILABLE:
+        throw new ServiceUnavailableException(Constants.Service.APP_FABRIC_HTTP, errMsg);
+    }
+    throw new IOException(
+      String.format("Failed to fetch artifact %s from app-fabric due to %s", artifactId, errMsg));
+  }
+
+  Path getLocalPath(String dirName, ArtifactId artifactId) {
+    return Paths.get(dataDir, dirName, artifactId.getNamespace(), artifactId.getArtifact(),
+                     artifactId.getVersion());
+  }
+
+  Path getLocalPath(String peerDir, String peerName, String dirName, ArtifactId artifactId) {
+      return Paths.get(dataDir, peerDir, peerName, dirName, artifactId.getNamespace(),
+                       artifactId.getArtifact(), artifactId.getVersion());
+  }
+
+  /**
+   * Returns a {@link File} representing the cache directory jars for the given artifact belonging to a tethered peer.
+   * The file path is:
+   * /DATA_DIRECTORY/peers/<peer-name>/</peer-name>artifacts/<namespace>/<artifact-name>/<artifact-version>/
+   */
+  private File getArtifactDirLocation(ArtifactId artifactId, String peerName) {
+    return getLocalPath("peers", peerName, "artifacts", artifactId).toFile();
+  }
+
+  /**
+   * Returns a {@link File} representing the cache directory jars for the given artifact. The file path is:
+   * /DATA_DIRECTORY/artifacts/<namespace>/<artifact-name>/<artifact-version>/
+   */
+  private File getArtifactDirLocation(ArtifactId artifactId) {
+    return getLocalPath("artifacts", artifactId).toFile();
+  }
+
+  /**
+   * Returns a {@link File} representing the cached jar for the given artifact and timestamp. The file path is:
+   * /DATA_DIRECTORY/peers/<peer-name></peer-name>/artifacts/<namespace>/<artifact-name>/<artifact-version>/
+   * <last-modified-timestamp>.jar if peerName is not null (i.e. artifact belongs to a tethered peer),
+   * /DATA_DIRECTORY/artifacts/<namespace>/<artifact-name>/<artifact-version>/<last-modified-timestamp>.jar otherwise
+   */
+  private File getArtifactJarLocation(ArtifactId artifactId, long lastModifiedTimestamp, @Nullable String peerName) {
```

Comment:
Since this is the base class, it shouldn't be aware of `peerName` at all. The base class should just be given an `artifactId`, call an abstract method to get the target `URL`, then fetch and cache it.

Reply:
Refactored code to remove references to peerName in the base class.